### ISSUE TITLE
Fill arrays with signaling NaNs upon object destruction

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#10822]
I'm reading "Code Complete" by Steve McConnell again with all of my students. It's a big book (nearly 1000 pages) but has much good advice, and one suggestion caught my eye: Filling memory with garbage when freeing it. We can use a specific kind of garbage: signaling NaNs. It might be interesting to try that out for our various vector and matrix classes.

**Comment by drwells** [comment]
This may be worth doing for things that allocate memory, but our modern compilers are good enough that they ignore writes to, e.g., `double values[dim]` since subsequent reads of that memory are undefined.

**Comment by rezarastak** [comment]
This approach tries to uncover hidden bugs, right? Could the same group of bugs be found using various sanitizers?

**Comment by bangerth** [comment]
Yes, it's for uncovering bugs. Sanitizers might have the same result, but with a larger effort: You have to re-compile things with a different compile flag, and there is probably a run-time overhead. Invalidating memory should have essentially no cost and aborts the program on first access of invalid memory.
